proxy/errors.go

```diff
@@ -0,0 +1,229 @@
+package proxy
+
+import (
+	"errors"
+	"io/ioutil"
+	"net/http"
+)
+
+type APIError struct {
+	Reason  StatusReason
+	Code    int
+	Status  string
+	Message string
+}
+
+func (e *APIError) Error() string {
+	return e.Message
+}
+
+const (
+	StatusSuccess = "Success"
+	StatusFailure = "Failure"
+)
+
+type StatusReason string
+
+const (
+	StatusReasonUnknown StatusReason = "Unknown"
+
+	// Status code 500
+	StatusReasonInternalError StatusReason = "InternalError"
+
+	// Status code 401
+	StatusReasonUnauthorized StatusReason = "Unauthorized"
```

Comment:
Could it be http.StatusUnauthorized? https://golang.org/src/net/http/status.go

Reply:
We can but we will have challenge in solving two problems, if we are not using `StatusReason`

1. When we will have same http status code being returned for different errors. I think we are not doing that right now but will be helpful in future. 

2. Since we are only having typed errors for only handful of http status codes, so if API server returns an HTTP response code for which we don't have typed error, we are returning that as `UnknownError` embedding HTTP Response code and reason. We are checking for UnknownError based on HTTP response code. Please go through this file, https://github.com/openfaas/faas-cli/pull/868/files#diff-25e38af31c2b62e171bb1b4e9a70546d19aa8fbd180d435435814962e2912600

If we remove StatusReason, to figure out UnknownError we will have to check for list of all HTTP status for which we have typed error.